File: java/S4_GatewayService/src/gatewayservice/GatewayService.java
/*
 *                         ADLINK Edge SDK
 *
 *   This software and documentation are Copyright 2018 to 2020 ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

/**
 * This code is part of example scenario 4 'Gateway Service' of the
 * ADLINK Edge SDK. For a description of this scenario see the
 * 'Edge SDK User Guide' in the /doc directory of the Edge SDK instalation.
 *
 * For instructions on building and running the example see the README
 * file in the Edge SDK installation directory.
 */

package gatewayservice;

import java.io.Closeable;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

//Use "AnsiConsole" from "jansi" library to print ANSI escape sequences both on Windows and Linux
import org.fusesource.jansi.AnsiConsole;
import org.fusesource.jansi.Ansi;
import org.fusesource.jansi.Ansi.Erase;

import com.adlinktech.datariver.*;

public class GatewayService implements Closeable {

    // This is a ConcurrentHashMap because it is used by a listener and listeners get triggered in different threads.
    public static Map<String, String> thingContext = new ConcurrentHashMap<String, String>();
    
    private static final int READ_DELAY = 10;
    private static final int TOTAL_HEADER_LINES = 2;
    private static final int TOTAL_FOOTER_MESSAGE_LINES = 1;
    private final Map<DataFlowKey, DataFlowValue> sampleCount;
    private final String thingPropertiesUri;
    private final DataRiver dataRiver;
    private final Thing thing;
    private final int screenHeightInLines;
    
    public GatewayService(final String thingPropertiesUri, final int screenHeight) {
        System.out.println("Gateway Service started");
        this.thingPropertiesUri = thingPropertiesUri;
        this.dataRiver = DataRiver.getInstance();
        this.thing = createThing();
        this.sampleCount = new HashMap<DataFlowKey, DataFlowValue>();
        this.screenHeightInLines = screenHeight;
    }
    
    private void cleanup() {
        if(this.dataRiver != null) {
            this.dataRiver.close();
        }
        System.out.println("Gateway Service stopped");
    }
    
    private String getAbsFileUri(final String fileName) {
        final String dir = System.getProperty("user.dir");
        final String prefix = "file://";
        return prefix.concat(dir).concat("/").concat(fileName);
    }

    private Thing createThing() {
        // Create and Populate the ThingClass registry with JSON resource files.
        final JSonThingClassRegistry tcr = new JSonThingClassRegistry();
        tcr.registerThingClassesFromUri(getAbsFileUri("definitions/ThingClass/com.adlinktech.example/GatewayServiceThingClass.json"));
        this.dataRiver.addThingClassRegistry(tcr);

        // Create a Thing based on properties specified in a JSON resource file.
        final JSonThingProperties tp = new JSonThingProperties();
        tp.readPropertiesFromUri(this.thingPropertiesUri);
        return this.dataRiver.createThing(tp);
    }
    
    private String truncate(final String str, final int width) {
        if (str.length() > width - 3) {
            return str.substring(0, width - 3) + "...";
        }

        return str;
    }

    private void displayStatus() {
        // Return if there is nothing to display
        if (this.sampleCount.entrySet().isEmpty()) {
            return;
        }

        // Move cursor position to the origin (0,0) of the console
        AnsiConsole.out.print(Ansi.ansi().cursor(0, 0));

        // Add header row for table
        displayHeader();

        // Write new data to console
        int lineCount = 0;
        for(final HashMap.Entry<DataFlowKey, DataFlowValue> data : this.sampleCount.entrySet()) {
            final DataFlowKey key = data.getKey();
            final DataFlowValue value = data.getValue();

            // Set grey color for purged flows
            final boolean alive = value.getFlowState() == FlowState.ALIVE;
            final Ansi COLOR1 = alive ? Ansi.ansi().fgGreen() : Ansi.ansi().fgDefault();
            final Ansi COLOR2 = alive ? Ansi.ansi().fgMagenta() : Ansi.ansi().fgBrightBlack();
            final String flowState = alive ? "" : " <purged>";

            AnsiConsole.out.println(
                COLOR1 + String.format("%-32s", truncate(key.getSourceThingContext() + flowState, 32))
                + Ansi.ansi().fgDefault()
                + COLOR2 + String.format("%-30s", truncate(key.getFlowId(), 30))
                + COLOR2 + String.format("%-20s", truncate(key.getTagGroupName(), 20))
                + Ansi.ansi().fgDefault()
                + Ansi.ansi().fgBrightBlack() + String.format("%-12s", truncate(key.getTagGroupQos(), 12))
                + Ansi.ansi().fgDefault()
                + COLOR1 + String.format("%8s", value.getSampleCount())
                + Ansi.ansi().fgDefault());

            lineCount++;

            if(lineCount < this.sampleCount.size() &&
                    lineCount >= (this.screenHeightInLines - TOTAL_HEADER_LINES - TOTAL_FOOTER_MESSAGE_LINES - 1)) {

                AnsiConsole.out.println(
                        "... " + (this.sampleCount.size() - lineCount) + " more lines available. "
                        + "Set terminal height to " + (this.sampleCount.size() + TOTAL_HEADER_LINES + TOTAL_FOOTER_MESSAGE_LINES + 1) + ". "
                        + "See the README file for more instructions.");
                break;
            }
        }
    }

    private void displayHeader() {
        System.out.println(
                String.format("%-32s", "Thing's ContextId")
                + String.format("%-30s", "Flow Id")
                + String.format("%-20s", "TagGroup Name")
                + String.format("%-12s", "QoS")
                + String.format("%8s", "Samples")
                + "\n");
    }

    private void readThingsFromRegistry() {
        final DiscoveredThingRegistry discoveredThingsRegistry = this.dataRiver.getDiscoveredThingRegistry();
        final DiscoveredThingSeq things = discoveredThingsRegistry.getDiscoveredThings();
        for (final DiscoveredThing thing : things) {
            thingContext.put(thing.getId(), thing.getContextId());
        }
    }
    
    public void run(final int runningTime) {
        final long startTimestamp = System.currentTimeMillis();
        long elapsedTime = 0;

        // Add listener for discovering new Things
        final NewThingDiscoveredListener newThingDiscoveredListener = new NewThingDiscoveredListener();
        this.dataRiver.addListener(newThingDiscoveredListener);

        // Get meta-data (contextId) for Things in discovered things registry
        readThingsFromRegistry();

        // Clear console screen before printing samples
        AnsiConsole.out.print(Ansi.ansi().eraseScreen(Erase.ALL));

        do {
            // Read data
            final IotNvpDataSampleSeq msgs = this.thing.readNextIotNvp("dynamicInput", (int)((runningTime * 1000) - elapsedTime));

            // Loop received samples and update counters
            for (final IotNvpDataSample msg : msgs) {
                final FlowState flowState = msg.getFlowState();

                final DataFlowKey key = new DataFlowKey(msg);
                
                if(!this.sampleCount.containsKey(key)) {
                    final DataFlowValue dfVal = new DataFlowValue();
                    this.sampleCount.put(key, dfVal);
                }

                // Store state in value for this flow
                this.sampleCount.get(key).setFlowState(flowState);

                // In case flow is alive or if flow is purged but sample
                // contains data: increase sample count
                final boolean sampleContainsData = (flowState == FlowState.ALIVE) || (msg.getData().size() > 0);

                if (sampleContainsData) {
                    this.sampleCount.get(key).setSampleCount(this.sampleCount.get(key).getSampleCount() + 1);

                    // In a real-world use-case you would have additional processing
                    // of the data received by msg.getData()
                }
            }

            // Update console output
            displayStatus();

            // Sleep before reading next samples
            try {
                Thread.sleep(READ_DELAY);
            } catch (InterruptedException e) {
                System.out.println("GatewayService is interrupted. Exiting");
                break;
            }

            // Get elapsed time
            elapsedTime = System.currentTimeMillis() - startTimestamp;
        } while (elapsedTime / 1000 < runningTime);

        // Remove listener
        this.dataRiver.removeListener(newThingDiscoveredListener);
    }

    @Override
    public void close() {
        cleanup();
    }

}

File: dotnet/S4_GatewayService/GatewayService/GatewayService.cs
/*                         ADLINK Edge SDK
 *
 *   This software and documentation are Copyright 2018 to 2020 ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

/**
 * This code is part of example scenario 4 'Gateway Service' of the
 * ADLINK Edge SDK. For a description of this scenario see the
 * 'Edge SDK User Guide' in the /doc directory of the Edge SDK instalation.
 *
 * For instructions on building and running the example see the README
 * file in the Edge SDK installation directory.
 */

using System;
using System.Collections;
using System.Collections.Generic;
using Adlinktech.DataRiver;

namespace GatewayService
{
    public static class Globals
    {
        public static IDictionary<string, string> ThingContexts = new Dictionary<string, string>();
    }

    public class DataFlowValue
    {
        public uint SampleCount { get; set; }
        public FlowState MyFlowState { get; set; }

        public DataFlowValue()
        {
            SampleCount = 0;
        }
    }

    public class DataFlowKey
    {
        public TagGroup MyTagGroup { get; set; }
        public string MySourceThingClassId { get; set; }
        public string MySourceThingId { get; set; }
        public string MyFlowId { get; set; }

        public DataFlowKey(IotNvpDataSample dataSample)
        {
            MyTagGroup = dataSample.TagGroup;
            MySourceThingClassId = dataSample.SourceClass;
            MySourceThingId = dataSample.SourceId;
            MyFlowId = dataSample.FlowId;
        }

        public string GetSourceThingContext()
        {
            string context = Globals.ThingContexts[MySourceThingId];
            if (String.IsNullOrEmpty(context))
            {
                context = "<unknown>";
            }
            return context;
        }
    }

    public class DataFlowKeyComparer : IEqualityComparer<DataFlowKey>
    {
        public bool Equals(DataFlowKey dataFlowKey1, DataFlowKey dataFlowKey2)
        {
            if (dataFlowKey1 == null && dataFlowKey2 == null)
                return true;
            else if (dataFlowKey1 == null || dataFlowKey2 == null)
                return false;
            else if (string.Equals(dataFlowKey1.MyTagGroup.Name, dataFlowKey2.MyTagGroup.Name)
                && string.Equals(dataFlowKey1.MySourceThingClassId, dataFlowKey2.MySourceThingClassId)
                && string.Equals(dataFlowKey1.MySourceThingId, dataFlowKey2.MySourceThingId)
                && string.Equals(dataFlowKey1.MyFlowId, dataFlowKey2.MyFlowId))
                return true;
            else
                return false;
        }

        public int GetHashCode(DataFlowKey dataFlowKey)
        {
            string fullId = dataFlowKey.MyTagGroup.Name
                + dataFlowKey.MySourceThingClassId
                + dataFlowKey.MySourceThingId
                + dataFlowKey.MyFlowId;
            return fullId.GetHashCode();
        }

    }

    //Listeners
    public class NewThingDiscoveredListener : ThingDiscoveredListener
    {
        public override void NotifyThingDiscovered(DiscoveredThing thing)
        {
            Globals.ThingContexts[thing.Id] = thing.ContextId;
        }
    }

    public class GatewayService : IDisposable
    {
        static DateTime StartTime;

        int ReadDelayMS = 10;
        int TotalHeaderLines = 2;
        int TotalFooterMessageLines = 1;

        public bool Stop { get; set; }
        private DataRiver DataRiver { get; }
        private Thing Thing { get; }
        private int WindowHeightInLines { get; }

        private IDictionary<DataFlowKey, DataFlowValue> SampleCount { get; set; }

        public GatewayService(string thingPropertiesURI, int windowHeight)
        {
            DataRiver = DataRiver.GetInstance();
            Thing = CreateThing(DataRiver, thingPropertiesURI);
            Stop = false;
            WindowHeightInLines = windowHeight;

            var comparer = new DataFlowKeyComparer();
            SampleCount = new Dictionary<DataFlowKey, DataFlowValue>(comparer);
            Console.CancelKeyPress += this.CtrlCHandler;
            Console.WriteLine("Gateway Service started");
        }

        public void Dispose()
        {
            DataRiver.Close();
            DataRiver.Dispose();
            Console.WriteLine("Gateway Service stopped");
        }

        private void CtrlCHandler(object sender, ConsoleCancelEventArgs args)
        {
            Console.WriteLine("Trapped Ctrl-C. Stopping.");
            Stop = true;
            args.Cancel = true; // don't terminate the process, we'll do that.
        }

        private static Thing CreateThing(in DataRiver dataRiver, string thingPropertiesURI)
        {
            // Create and Populate the ThingClass registry with JSON resource files.
            var tcr = new JSonThingClassRegistry();
            tcr.RegisterThingClassesFromUri("file://./definitions/ThingClass/com.adlinktech.example/GatewayServiceThingClass.json");
            dataRiver.AddThingClassRegistry(tcr);

            // Create a Thing based on properties specified in a JSON resource file.
            var tp = new JSonThingProperties();
            tp.ReadPropertiesFromUri(thingPropertiesURI);

            return dataRiver.CreateThing(tp);
        }

        public int Run(int runningTime)
        {
            StartTime = DateTime.Now;

            // Add listener for discovering new Things
            var newThingDiscoveredListener = new NewThingDiscoveredListener();
            DataRiver.AddListener(newThingDiscoveredListener);

            // Get meta-data (contextId) for Things in discovered things registry
            ReadThingsFromRegistry();

            // Clear console window before printing samples
            Console.Clear();

            do
            {
                // Read data
                int timeout = (int)(runningTime - SecondsExecuting(StartTime)) * 1000;
                var msgs = Thing.ReadNextIotNvp("dynamicInput", timeout);

                // Loop received samples and update counters
                foreach (var msg in msgs)
                {
                    var flowState = msg.FlowState;

                    // Store state in value for this flow
                    var key = new DataFlowKey(msg);
                    if (!SampleCount.ContainsKey(key))
                    {
                        SampleCount.Add(key, new DataFlowValue());
                    }
                    var v = SampleCount[key];
                    v.MyFlowState = flowState;

                    // In case flow is alive or if flow is purged but sample
                    // contains data: increase sample count
                    bool sampleContainsData = (flowState == FlowState.ALIVE) || (msg.Data.Count > 0);

                    if (sampleContainsData)
                    {
                        v.SampleCount++;

                        // In a real-world use-case you would have additional processing
                        // of the data received by msg.getData()
                    }
                }

                // Update console output
                DisplayStatus();

                // Sleep before reading next samples
                System.Threading.Thread.Sleep(ReadDelayMS);

            } while (!Stop && SecondsExecuting(StartTime) < runningTime);

            // Remove listener
            DataRiver.RemoveListener(newThingDiscoveredListener);

            return 0;
        }

        private double SecondsExecuting(DateTime originalTime)
        {
            return (DateTime.Now - originalTime).TotalSeconds;
        }

        private void ReadThingsFromRegistry()
        {
            var discoveredThingsRegistry = DataRiver.DiscoveredThingRegistry;
            var things = discoveredThingsRegistry.DiscoveredThings;
            foreach (var t in things)
            {
                Globals.ThingContexts[t.Id] = t.ContextId;
            }
        }

        private void DisplayStatus()
        {
            // Move cursor position to the origin (0,0) of the console
            Console.SetCursorPosition(0, 0);

            // Add header row for table
            DisplayHeader();

            // Write new data to console
            int lineCount = 0;

            foreach (var item in SampleCount)
            {
                DataFlowKey key = item.Key;
                DataFlowValue v = item.Value;

                // Set grey color for purged flows
                bool alive = v.MyFlowState == FlowState.ALIVE;
                var color1 = alive ? ConsoleColor.DarkGreen : Console.ForegroundColor;
                var color2 = alive ? ConsoleColor.Magenta : ConsoleColor.DarkGray;
                string flowStateStatus = alive ? "" : " <purged>";

                Console.ForegroundColor = color1;
                Console.Write("{0,-32}", Truncate(key.GetSourceThingContext() + flowStateStatus, 32));
                Console.ForegroundColor = color2;
                Console.Write("{0,-30}", Truncate(key.MyFlowId, 30));
                Console.ForegroundColor = color2;
                Console.Write("{0,-20}", Truncate(key.MyTagGroup.Name, 20));
                Console.ForegroundColor = ConsoleColor.DarkGray;
                Console.Write("{0,-12}", Truncate(key.MyTagGroup.QosProfile, 12));
                Console.ForegroundColor = color1;
                Console.WriteLine("{0,8}", v.SampleCount);
                Console.ResetColor();

                lineCount++;

                if (lineCount < SampleCount.Count &&
                    lineCount >= (WindowHeightInLines - TotalHeaderLines - TotalFooterMessageLines - 1))
                {
                    Console.Write($"... {SampleCount.Count - lineCount} more lines available. ");
                    Console.Write($"Set terminal height to {SampleCount.Count + TotalHeaderLines + TotalFooterMessageLines + 1}. ");
                    Console.WriteLine("See the README file for more instructions.");
                    break;
                }
            }

        }

        private void DisplayHeader()
        {
            Console.WriteLine("{0,-32}{1,-30}{2,-20}{3,-12}{4,8}\n",
              "Thing's ContextId",
              "Flow Id",
              "TagGroup Name",
              "QoS",
              "Samples"
          );

        }

        private string Truncate(string source, int length)
        {
            if (source.Length > length)
            {
                source = source.Substring(0, length);
            }
            return source;
        }
    }
}
